### DISCUSSION TITLE
Mocking an output port on UNIT TEST

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#1414] [Q&A] [answered]
Does Fprime provide a way to mock an output port on Unit tests? If yes, how?

I'll try to explain with a simple example:

> Let's consider two components: `Producer` and `Consumer`.
> 
> -  `Producer` has an output port called `provideLastState`
> -  `Consumer` has an input port called `getLastState`
> 
> The two-port are connected together. When Consumer calls the port `getLastState`, the `Consumer` put the last state inside a variable passed by reference called `my_state`. What I would like to do is to create a mock/fake of `getLastState`so in the unit test of `Consumer` I can set the `my_state` as I prefer and test different behaviour of the `Consumer` based on the value of `my_state`.

**Comment by LeStarch** [comment, marked as answer]
Yes we do!  https://nasa.github.io/fprime/UsersGuide/user/unit-testing.html

The unit test harness created by `fprime-util impl --ut` has stubbed port implementations for all ports on a component.  Thus you can write whatever remote behavior you want and test the component in isolation.

In this example you'd create a test harness for Consumer and implement your mock state production in the test harness. 

If you need more details to get started, please post back here!

**Reply by SMorettini** [reply]
Thank you for the answer. Could you give me some more details? I already tried to use the functions provided by TesterBase but I didn't find a way to solve the problem yet.

Should I implement my mock inside the following autogenerated function in `Tester.cpp`? Or I'm approaching the problem in a total wrong way?
```
  // ----------------------------------------------------------------------
  // Handlers for typed from ports
  // ----------------------------------------------------------------------

  void Tester ::
    from_getLastState_handler(
        const NATIVE_INT_TYPE portNum,
        Type::State state
    )
  {
    this->pushFromPortEntry_getLastState(state);
  }
```

**Reply by codeflight1** [reply]
You are approaching it the wrong way, you need to write a test in the Tester::toDo() function, and you can the use ASSERT_FROM_<portName> to check the output. It might help to take a look at the MathComponent tutorial.

**Reply by kevin-f-ortega** [reply]
> Should I implement my mock inside the following autogenerated function in Tester.cpp?  

@SMorettini yes. If you want to mock the output port's behavior, then write it out inside the `from_` autogenerated function in Tester.cpp.

**Reply by timcanham** [reply]
It also isn't uncommon if you need to implement special code for testing to augment what is in the handler. You can add your own code and even comment out the `push()` statement if you don't care. `Tester.cpp` was meant to be initially generated as a stub that the developer fills in. I frequently set up a condition in `Tester.cpp`, invoke the command or port that in turn invokes the handler, and then check the outcome. For instance, in your example (I modified the signature to have a reference to modify the state):

```c++
  // ----------------------------------------------------------------------
  // Handlers for typed from ports
  // ----------------------------------------------------------------------

  void Tester ::
    from_getLastState_handler(
        const NATIVE_INT_TYPE portNum,
        Type::State &state
    )
  {
    state = this->testState;
    this->pushFromPortEntry_getLastState(state);
  }
```
In the test case, you can do:

```c++
 ...
 this->testState = MY_STATE;
 this->invoke_to_someport(...); // this causes getLastState() port to be called
 ASSERT_TRUE(something that happened because state == MY_STATE)
```
